### PR TITLE
Feature + Tests: i64 operations

### PR DESCRIPTION
**Pull request by nerodesu017**
### Pull Request Overview

This pull request adds support for i64 operations.

### Testing Strategy

This pull request was tested locally using the provided test cases.

### TODO or Help Wanted

This pull request still needs reviews over the refactoring of the `WasmReader` structure. This was modified because I have started working on it based on the branch related to the magic bytes converted to slices (#23).

### Formatting

- [x] Ran `cargo fmt`
- [x] Ran `cargo check`
- [x] Ran `cargo build`
- [ ] Ran `nix fmt`
- [ ] Ran `treefmt`

### Github Issue

This pull request partially implements #2.

### Author

Signed-off-by: Popescu Adrian <adrian.popescu@oxidos.io>

### CODE REVIEW

File: src/core/reader/types/values.rs [hunk 1]
```diff
@@ -14,14 +14,19 @@ use crate::{Error, Result};
 impl WasmReader<'_> {
     /// Note: If `Err`, the [WasmReader] object is no longer guaranteed to be in a valid state
     pub fn read_u8(&mut self) -> Result<u8> {
-        let value = *self.current.first().ok_or(Error::Eof)?;
+        if self.full_contents.len() <= self.pc {
```
Comment (valexandru) [inline src/core/reader/types/values.rs:17]: You can use the peek_u8() to get the value and check if the pc is valid at the same time

File: src/execution/mod.rs [hunk 2]
```diff
@@ -328,6 +333,191 @@ where
...
+                    }
+
+                    let res = divisor.checked_rem(dividend);
+                    let res = res.unwrap_or_default();
+
+                    trace!("Instruction: i64.rem_s [{divisor} {dividend}] -> [{res}]");
+                    stack.push_value(res.into());
+                }
+                I64_REM_U => {
+                    // this is a correct implementation
```
Comment (valexandru) [inline src/execution/mod.rs:430]: I would rather have the reasoning in the comment than an implementation link that does the same

File: tests/arithmetic/bitwise.rs [hunk 3]
```diff
@@ -44,11 +50,9 @@ pub fn i32_bitwise_or() {
     assert_eq!(-1, instance.invoke_func(0, (i32::MIN, i32::MAX)).unwrap());
 }
 
-/// A simple function to test the i32.xor bitwise operation
+// /// A simple function to test the i32.xor bitwise operation
```
Comment (valexandru) [inline tests/arithmetic/bitwise.rs:53]: Possible typo here?

Comment (valexandru) [reply tests/arithmetic/bitwise.rs:53]: Also appears like this in multiple locations

File: src/execution/mod.rs (hunk 2 above)
Comment (nerodesu017) [reply src/execution/mod.rs]: I removed it altogether, I assumed it was too trivial to have it.

File: tests/arithmetic/remainder.rs [hunk 4]
```diff
@@ -1,27 +1,167 @@
...
-    use wasm::{validate, RuntimeInstance};
+pub fn i64_remainder_signed_simple() {
+    let wat = String::from(REM_S_WAT).replace("{{TYPE}}", "i64");
+    let wasm_bytes = wat::parse_str(wat).unwrap();
+    let validation_info = validate(&wasm_bytes).expect("validation failed");
+    let mut instance = RuntimeInstance::new(&validation_info).expect("instantiation failed");
 
-    let wat = REM_S_WAT;
+    assert_eq!(
+        0 as i64,
```
Comment (george-cosma) [inline tests/arithmetic/remainder.rs:29]: I think you can use `0_i64`, `999_i64` etc. In any case, lgtm

**Review by george-cosma — APPROVED** [review]

**Comment by florianhartung** [comment]
Otherwise this looks good to me, good job @nerodesu017!
When the conflicts (due to the merge of https://github.com/DLR-FT/wasm-interpreter/pull/41) are resolved, we are ready for a merge👍

**Comment by wucke13** [comment]
I inferred that only the `feat: i64 ops` is the relevant one. I cherry picked that one onto a fresh main branch over https://github.com/DLR-FT/wasm-interpreter/tree/feature/i64-ops-wucke13-fix to solve the merge conflict. Please take a look at it, if you like it, I will force-push that over this PR's branch so that we can eventually merge this PR.

**Review by george-cosma — APPROVED** [review]

**Review by wucke13 — APPROVED** [review]

**Pull request merged**